### DISCUSSION TITLE
single machine 6 HDD - chunkservers [question]

### DISCUSSION THREAD

**Discussion by eleaner** [moosefs/moosefs#630] [Q&A]
Hi

I have an elderly AMD Turion II Neo N54L Dual-Core with 6HDD.
I have set it up with one chunkserver per each HDD but as you can imagine my load is often exceeding 16.
Would it be safe/correct to use one chunk server for all HDDs?
Technically making it moosefs with single chunkserver

**Comment by marcomilano** [comment]
I don't understand your setup, are you using one physical server for the entire moosefs?

**Comment by eleaner** [comment]
it is not exactly this way, but close enough.
It's a home setup. I have two 6HDD file servers. I intend to run them together but after I finished my tests. So right now I have only one available
I am supplementing it with 11TB on my workstation but since my fileserver has 40TB it's almost like it's running alone.

Enough to say I have another home machine with 15 drives of different sizes and I am considering running moose pro with ec:8+2 on it (assuming it can handle the load)

So the short answer is yes.
It's HDD failure protection at the moment rather than server failure protection.

Answering my own question
one chunkserver equals one chunk
so I can easily run one chunkserver on all six drives but I will need a second machine if I want any resiliency at all.
At the moment I settled on two chunk servers for my 6HDD and divided the drive capacity equally between them.

Ultimately one chunkserver for one HDD would be the best solution, but I just don't have enough CPU.

Just a word of caution.
When I hooked up all six drives to one chunkserver it realised that it has a lot of additional chunks (I had a goal of 2). And since it cannot hold more than one chunk it dumped all the second copies in virtually no time at all.
So right now I have to wait until it rebalances again.

**Comment by marcomilano** [comment]
> it is not exactly this way, but close enough.
> It's a home setup. I have two 6HDD file servers. I intend to run them together but after I finished my tests. So right now I have only one available
> I am supplementing it with 11TB on my workstation but since my fileserver has 40TB it's almost like it's running alone.
> 
> Enough to say I have another home machine with 15 drives of different sizes and I am considering running moose pro with ec:8+2 on it (assuming it can handle the load)
> 
> So the short answer is yes.
> It's HDD failure protection at the moment rather than server failure protection.
> 
> Answering my own question
> one chunkserver equals one chunk
> so I can easily run one chunkserver on all six drives but I will need a second machine if I want any resiliency at all.
> At the moment I settled on two chunk servers for my 6HDD and divided the drive capacity equally between them.
> 
> Ultimately one chunkserver for one HDD would be the best solution, but I just don't have enough CPU.
> 
> Just a word of caution.
> When I hooked up all six drives to one chunkserver it realised that it has a lot of additional chunks (I had a goal of 2). And since it cannot hold more than one chunk it dumped all the second copies in virtually no time at all.
> So right now I have to wait until it rebalances again.

OK, at least you are aware that your setup of moosefs is somewhat unusual.
Normally people will use zfs for single box data redundancy and leave MooseFS
for truely distributed SDS setups, but anyways it is a free world, you can use it
any way you want as long as you are aware that your usage pattern is not really
what this is designed for.

**Comment by eleaner** [comment]
growing space in zfs is a pain in the butt
growing moose is just replacing hdd
and moose can handle different hdd sizes easily
since these are some of my top priorities I fail to see too many benefits of zfs here

**Comment by marcomilano** [comment]
I love MooseFS but if I had only one box , I would simply use zfs on it.
It is a personal choice, if you want to use moosefs on a single box, then go ahead and use it,
it is designed to be a networked, distributed SDS, but yes one can use it like the way you are using it,
your usage pattern is just not a mainstream use case.

**Comment by nickb937** [comment]
I did something similar.  I had two machines with significantly differing HDD capacities such that it was not feasible to have a filesystem of sufficient capacity and 100% machine resilience.  I ran a chunkserver per HDD mount rather than a single chunkserver spanning multiple mounts. 

Generally this setup worked well for a few years.  The data was well balanced across all HDDs and the system was resilient to a couple of HDDs failing (I'd previously experienced double-disk loss on a RAID5 array due to buying five HDDs from the same batch).  Upgrading individual HDDs as they become end-of-life was easy - they either automatically drained because they broke or I manually marked them for draining, then swapped out the drive with a much larger drive (e.g. 2TB to 6TB), left the system 1-2 days to rebalance itself and enjoyed the growth in filesystem space too.

The system was not resilient to a whole machine failing, but that wasn't necessarily my objective.  However a local lightning strike did fry the on-board NIC on one machine, leaving me a degraded system  with a bunch of files missing chunks for a while.

Lately I've switched to a single chunkserver per machine, so each chunkserver has control over several HDDs, for no particular reason other than slightly improved data resiliency which I could have implemented using the other setup by changing the goals.

You could use ZFS on a single box, if you have only a single box, but it depends on your long-term goals. 
 If long-term you're looking to have multiple servers, then setting up MooseFS initially on a single server will be quite reasonable.  Setting up the second server in the future then becomes a doddle: configure the chunkserver - job done.

**Comment by eleaner** [comment]
Thank you @nickb937 
I used to run ZFS but it is expensive to grow especially with 6 hdd.
resilience wise I was running raid 10 anyway so it's neither loss nor gain space wise to MooseFS replication.
I have two machines at home (although the migration will take time) and longer term I think of setting two chunkservers to allow one machine failure.
as a benefit, I should also have a read boosts when accessing the data

**Comment by VideoFX** [comment]
Im using ZFS + MooseFS for homelab. I think that's the way to do it. 
Some observations:
- It is very slow to do it like this, but works amazing! 
- If you pair equal size disks in zfs raid 0, it speeds right up! (what, raid 0? yes, live a little, plus moosefs does the redundancy) 
- Add as many chunkservers as you have raid pools, or if you have a lot of raid pools, add as many as needed for redundancy.
- Viola! 
- This can be done on a single machine! Just add more machines and/or zfs drives/pools as needed (more than one machine would be optimal ofc).
- Between mfscache and zfs's ARC cache, this can get really fast!
- zfs's built in compression also squeezes out a significant amount of capacity it seems. I was impressed immediately with zfs + moosefs.
- Build one of these and tell your friends, they will look at you with a long blank stare, and probably never understand, but you will feel like the first person to set foot on the moon.

Ok, so why not just use zfs raidz and skip moosfs altogether for homelab? Moosefs adds so many features to a raid. It would crazy not to add moosfs to one's zfs raid (imo). <3

**Comment by eleaner** [comment]
what about the flexibility of scaling your solution?

for me - one of the massive benefits of moosefs is that I can pair together
any size of the HDD at any time (just need to balance sizes of chunk
servers).
In my case it is not homelab - it is mainly photo storage and archive, so I
need to keep growing it.

If I remember growing the ZFS pools is a bit nightmarish

on archive/backup server I paired moosefs with S3QL for space efficiency
But I am not concerned with performance too much there

**Comment by VideoFX** [comment]
Yes, same for me with ZFS. Thats why I love MooseFS, just add more drives and/or machines. You can just grow and grow. The chunkservers and drives dont even have to be the same size, the only time the size must match is when I do a zfs raid 0. 

Quick note about zfs raids + moosefs, my raids are small, like 2 or 3 drive raids, thats it. I remember reading that large raids are not suggested for moosefs, and that makes total sense. However with small raids, it speeds up moosefs quite a bit for me. Plus, if one of the raids breaks, I can reformat/resilver/rebuild the raid, it doesn't take a week (maybe a few hours), and the other drives can compensate for the lost raid until it's back up. Strange design, but seems to work nicely.

I just built a 100TB moosefs system today in my lab. I had drives of all sizes, 4, 6, 8TB, some of them were even SMR drives (yikes).  Big mix of drives in two servers, close to 50TB each server. Multiple chunkservers within each server. The reason there is multiple is for flexibility and drive labeling. Also, I will often use a separate chunkserver for SSDs too, so there will usually be some SSDs and NVMEs in there too.

Oh, quick note on SMR drives. I didn't believe people when they said SMR drives sucked (I have some WD blues, and the old SMR WD Reds). It's true, in mooseFS, with zfs, they are slow. Good golly, but again, when in raid 0, they speed right up. If you are the sad owner of some SMR drives and you are feeling the low performance. Put them in a performance raid and they come alive. 

Oh, another note on SMR drives with zfs and moosfs: Apparently SMR drives are slow and can lead to errors reported in zfs, which will also lead to a error reported in the moosfs GUI. This can happen because SMR drives are so darn slow (maybe someone can explain it better, but I can summarize here), and I will see errors once and a while on heavy use. This can lead users to think the drives are going bad or have bad sectors. However after testing the drives they seem fine, and after putting those SMR drives in raid 0 I have not seen an error. I think smr drives may be sensitive to that in zfs. There were some other zfs users mentioning they have seen this with smr drives as well.

Anyway, SMR drives are not significantly cheaper like they use to be, so I also suggest not buying them if you are in the market for more drives. If you already have them, throw them in a moosefx box ofc!

**Comment by rinkydinkyproject** [comment]
> I did something similar. I had two machines with significantly differing HDD capacities such that it was not feasible to have a filesystem of sufficient capacity and 100% machine resilience. I ran a chunkserver per HDD mount rather than a single chunkserver spanning multiple mounts.
> 
> Generally this setup worked well for a few years. The data was well balanced across all HDDs and the system was resilient to a couple of HDDs failing (I'd previously experienced double-disk loss on a RAID5 array due to buying five HDDs from the same batch). Upgrading individual HDDs as they become end-of-life was easy - they either automatically drained because they broke or I manually marked them for draining, then swapped out the drive with a much larger drive (e.g. 2TB to 6TB), left the system 1-2 days to rebalance itself and enjoyed the growth in filesystem space too.
> 
> The system was not resilient to a whole machine failing, but that wasn't necessarily my objective. However a local lightning strike did fry the on-board NIC on one machine, leaving me a degraded system with a bunch of files missing chunks for a while.
> 
> Lately I've switched to a single chunkserver per machine, so each chunkserver has control over several HDDs, for no particular reason other than slightly improved data resiliency which I could have implemented using the other setup by changing the goals.
> 
> You could use ZFS on a single box, if you have only a single box, but it depends on your long-term goals. If long-term you're looking to have multiple servers, then setting up MooseFS initially on a single server will be quite reasonable. Setting up the second server in the future then becomes a doddle: configure the chunkserver - job done.

Sorry to necro this, BUT
I was unable to find answers to this question, I'm trying to setup something similar, but how did you make each mount a chunkserver in the configs? Also did you use existing data when setting it up (is that even possible)? I'm looking to migrate away from an existing solution and have a way to do that, BUT I've seen no clear answer on whether or not I can use prepopulated drives on Moosefs without losing data (I do have backups if required but would prefer not to restore if I don't have to). Right now I'm setting up a single system instance of MooseFS (and I recognize it's not designed for this but clearly people use it this way) so I'm looking to do this for now until I can expand into second/third chunkservers. BUT for now happy to just run the one and have taken steps to keep the data safe. JUST so long as I know I can bring on existing data with minimal/no risk?

**Comment by VideoFX** [comment]
Not sure if this answers it, but you can launch multiple chunkservers on a single machine. So if you want to, you could make a separate config for each chunkserver, and each of those configs may use a different drive. This is how you run multiple chunkservers on the same machine, multiple configs, and multiple instances of the chunkserver for each config.

For existing data, I don't think its possible (at least not that I know of). Moosefs will not overwrite your existing data, but it also won't import it (I think, correct me if I'm wrong, maybe I just didn't try this). So the existing data on the drive will be untouched by the chunkserver. If you want to import data so that its part of the moosefs volume, you must copy the data to the moosefs directory. Then it will be part of the moosefs database and follow whatever replication rules you have set for that directory. I hope that made sense.

I still run multiple chunk servers on a single machine, after all these years. I think its an excellent use case, and I also think its within the design of moosefs, although its not recommended. It really demonstrates the power of the platform. You can think of replication between singles drives WITHIN a single machine, or BETWEEN multiple machines.. or both! I think I have a total of 8 chunk servers running between two servers. The redundancy both within and between machines is agreeable for my needs. I do this instead of using raidz (although I still use zfs, and even zfs raid 0 on some drive pairs), and find management of drives is easier.

**Reply by chogata** [reply]
This is not true - MooseFS can use chunks on existing drives in a new chunkserver process, as long as you just move the drives within the same MooseFS instance. It won't use a drive with chunks from another instance (makes sense - so you don't accidentally  move a drive between instances and totally mess up your data), but within the same instance it's possible. There are some sanity rules to obey, a rule of thumb is: don't put drives that might contain a copy of the same chunk in the same chunkserver (same process).
So it's entirely possible to take a MooseFS instance that has, let's say, 1 master process and 2 chunk server processes with 3 drives each and convert it into an instance with the same master process(!) and 6 chunk server processes, 1 drive each, using the drives with data "as they are".

**Comment by rinkydinkyproject** [comment]
THANK YOU! Yes this makes a bit more sense. I just need to sort out the configuration part for that. It's still fairly new for me. As I'm not running a production/enterprise setup, not overly concerned about it not being recommended so long as it can do it. My main reasons for switching over are the at a glance monitoring tools PLUS the ability to query the space as WELL as the balancing/replication (which appears it can do just fine). I'm using BTRFS for my drives but no raid. The recovery tools+snapshots PLUS offsite backups have me covered. Was using a uniontool but the performance of it was problematic plus lack of alerting & balancing which I need working well. My goal is eventually to build it into a cluster using my other machines as chunkservers, BUT for now I just need a single machine setup. 

SO if I understand correctly, I have to create a chunkserver config for EACH drive? What does this look like exactly? Do I create an mfschunkserver1.cfg mfschunkserver2.cfg, etc? I appreciate the reply. This is helpful for me and I'm sure others out there trying this out (and from what I've seen there are a few out there using MooseFS). Knowing the data won't be affected, just won't be imported is a huge deal for me, because it means once I sort out the chunkserver piece I can start it then just move the data over into the MFSpool.

**Comment by Lindsay-Mathieson** [comment]
> SO if I understand correctly, I have to create a chunkserver config for EACH drive?

You can have multiple drives per chunkserver, or just one if you want. Its best to have the total storage per chunkserver to be roughly the same (but not essential).

Replication is per chunkserver, not per drive. i.e, if your want RAID1 (Copies = 2), then you need a minimum of two chunkservers.

When starting a chunkserver, you can specify its config file via cmd line args, thats how you assigned different drives to them. You'll need to create custom systemd scripts for them as well.

You might find it easier to run the chunkserver's via lxc, thats what I do. That way each chunkserver has its own virtual machice (lxc) and you don't have to findle with command line scripts.

**Comment by rinkydinkyproject** [comment]
Interesting setup with LXC to consider and certainly an option I'll explore if it's not too resource intensive for me and I'm familiar enough with containers for this sort of things given I've built kubernetes clusters via terraform and such (tho at home I do video editing, photography and other resource intensive things with my system so containers and VMs can rob my system of needed resources for that work). Tho I am hoping for @VideoFX setup on this one given he's been running it for years this way.
